# Importing the Libraries

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# Importing The Dataset

In [2]:
df= pd.read_csv("hw1_q2_dataset.csv",header = None)
X = df.drop([4],axis=1)
X = np.array(X)
Y = df[4]
Y = np.array(Y)

# Implementing Agglomerative Clustering With Single Linkage and Complete Linkage

In [3]:
def HAC(X,method):
    u = method.lower()
    if u == "single":
        h = singlelink(X)
        return h
    elif u == "complete":
        h = complink(X)
        return h

#Implementing single linkage clustering 

def singlelink(X):
    matrix = []
    slink = []

    for i in range(0,len(X)):
        diff = X-X[i]
        diff = np.square(diff)
        diff = np.sum(diff,axis=1)
        diff = np.sqrt(diff)
        matrix.append(diff)
    matrix = np.array(matrix)
    x = matrix
    rows = []
    for i in range(0,len(X)):
        cols=[]
        cols.append(i)
        rows.append(cols)
    n = rows
    v = len(n)
    while(v!=1):   
        c=100000
        for i in range(0,len(x)):
            for j in range(0,len(x[i])):
                if x[i][j]<=c:
                    if x[i][j]==0:
                        c=c
                    else:
                        c=x[i][j]
                        a=i
                        b=j
                    
        li = []
        for i in range(0,len(n[a])):
            li.append(n[a][i])
        for j in range(0,len(n[b])):
            li.append(n[b][j])
        n[a] = None
        n[a] = li
        n[b]=None
        g = list(n)
        while(None in g):
            g.remove(None)
        for i in range(0,len(x[a])):
            if x[a][i]>=x[b][i]:
                x[a][i]=x[b][i]
        for i in range(0,len(x)):
            if x[i][a]>=x[i][b]:
                x[i][a]=x[i][b]
        for i in range(0,len(x[b])):
            x[b][i]=0
        for i in range(0,len(x)):
            x[i][b]=0
        v = len(g)
        slink.append(g)
    return slink

#Implementing complete linkage

def complink(X):
    matrix = []
    slink = []
    for i in range(0,len(X)):
        diff = X-X[i]
        diff = np.square(diff)
        diff = np.sum(diff,axis=1)
        diff = np.sqrt(diff)
        matrix.append(diff)
    matrix = np.array(matrix)
    x = matrix
    rows = []
    for i in range(0,len(X)):
        cols=[]
        cols.append(i)
        rows.append(cols)
    n = rows
    v = len(n)
    while(v!=1):   
        c=100000
        for i in range(0,len(x)):
            for j in range(0,len(x[i])):
                if x[i][j]<=c:
                    if x[i][j]==0:
                        c=c
                    else:
                        c=x[i][j]
                        a=i
                        b=j
                    
        li = []
        for i in range(0,len(n[a])):
            li.append(n[a][i])
        for j in range(0,len(n[b])):
            li.append(n[b][j])
        n[a] = None
        n[a] = li
        n[b]=None
        g = list(n)
        while(None in g):
            g.remove(None)
        for i in range(0,len(x[a])):
            if x[a][i]<=x[b][i]:
                if i==a or i==b:
                    x[a][i]=0
                    x[b][i]=0
                else:
                    x[a][i]=x[b][i]
        for i in range(0,len(x)):
            if x[i][a]<=x[i][b]:
                if i==a or i==b:
                    x[i][a]=0
                    x[i][b]=0
                else:
                    x[i][a]=x[i][b]
        for i in range(0,len(x[b])):
            x[b][i]=0
        for i in range(0,len(x)):
            x[i][b]=0
        v = len(g)
        slink.append(g)
    return slink

        
#Implementing the conversion of labels to integer   
    
def trltn(label):
    cols = []
    for i in range(0,len(label)):
        if label[i] == ' Metal':
                cols.append(0)
        elif label[i] == ' Ceramic':
                cols.append(1)
        elif label[i] == ' Plastic':
                cols.append(2)
    return cols
       

#Implementation of weighted accuracy and each cluster accuracy seprately    

def accuracy(X,Y,n):
    
    wc = 0
    acc = []
    ans = X[len(X)-n]
    label = []
    for i in range(0,len(ans)):
        col = []
        for j in range(0,len(ans[i])):
            n = ans[i][j]
            col.append(Y[n])
        label.append(col)
    for i in range(0,len(label)):
        c = 0
        p = np.bincount(label[i]).argmax()
        for j in range(0,len(label[i])):
            b = ans[i][j]
            if Y[b] == p:
                c=c+1
        c = c/len(ans[i])
        c=c*100
        acc.append(c)
        c = c*len(ans[i])
        c = c/len(Y)
        wc = wc+c
    acc.append(wc)
    return acc
    

# Converting The Labels Into Integer Value

In [4]:
Y_train = trltn(Y)

# Implementing Hierachical Clustering On The Dataset

In [14]:
s = HAC(X,"single")
c = HAC(X,"complete")

# Accuracy OF Every Cluster And Weighted Accuracy For Following Number Of Cluseters for Single Linkage As Follows

In [29]:
for i in [2,4,6,8]:
    print("Accuracy for " +str(i)+" Cluster clusterwise is \n")
    acc = accuracy(s,Y_train,i)
    print(str(acc[:-1])+"\n")
    print("Weighted Accuracy for " +str(i)+" Cluster is \n")
    print(str(acc[-1])+"\n")

Accuracy for 2 Cluster clusterwise is 

[100.0, 50.42016806722689]

Weighted Accuracy for 2 Cluster is 

50.833333333333336

Accuracy for 4 Cluster clusterwise is 

[100.0, 100.0, 100.0, 50.427350427350426]

Weighted Accuracy for 4 Cluster is 

51.666666666666664

Accuracy for 6 Cluster clusterwise is 

[100.0, 100.0, 100.0, 100.0, 50.0, 51.81818181818182]

Weighted Accuracy for 6 Cluster is 

53.333333333333336

Accuracy for 8 Cluster clusterwise is 

[100.0, 75.0, 100.0, 100.0, 100.0, 100.0, 100.0, 52.293577981651374]

Weighted Accuracy for 8 Cluster is 

55.83333333333333



# Accuracy OF Every Cluster And Weighted Accuracy For Following Number Of Cluseters for Single Linkage As Follows

In [28]:
for i in [2,4,6,8]:
    print("Accuracy for " +str(i)+" Cluster clusterwise is \n")
    acc = accuracy(c,Y_train,i)
    print(str(acc[:-1])+"\n")
    print("Weighted Accuracy for " +str(i)+" Cluster is \n")
    print(str(acc[-1])+"\n")

Accuracy for 2 Cluster clusterwise is 

[50.0, 50.0]

Weighted Accuracy for 2 Cluster is 

50.0

Accuracy for 4 Cluster clusterwise is 

[58.82352941176471, 40.0, 50.0, 50.0]

Weighted Accuracy for 4 Cluster is 

50.0

Accuracy for 6 Cluster clusterwise is 

[62.06896551724138, 58.82352941176471, 40.0, 37.93103448275862, 40.0, 70.0]

Weighted Accuracy for 6 Cluster is 

50.0

Accuracy for 8 Cluster clusterwise is 

[100.0, 100.0, 62.06896551724138, 58.82352941176471, 50.0, 37.93103448275862, 40.0, 77.77777777777779]

Weighted Accuracy for 8 Cluster is 

53.33333333333333



# Results And Observation

### As we can see from above obtained results that weighted accuracy for total number of 8 cluters is highest amongst both single linkage clustering and complete linkage clustering but comparing both we can say that single linkage clustering slightly edges onto the complete linkage clustering as both of the clustering techniques have weighted accuracy of 55.83% and 53.33% respectievely

### Accuracy of single linkage > Accuracy of complete linkage